tower-filter/src/error.rs

```diff
@@ -0,0 +1,54 @@
+//! Error types
+
+use std::error;
+use std::fmt;
+
+/// Error produced by `Filter`
+#[derive(Debug)]
+pub struct Error {
+    source: Option<Source>,
+}
+
+pub(crate) type Source = Box<dyn error::Error + Send + Sync>;
+
+impl Error {
+    /// Create a new `Error` representing a rejected request.
+    pub fn rejected() -> Error {
+        Error { source: None }
+    }
+
+    /// Create a new `Error` representing an inner service error.
+    pub fn inner<E>(source: E) -> Error
```

Comment:
It seems one can create an "inner" error even in a predicate.

Reply:
This is by design. The predicate can error w/o actually rejecting. In this case, it is represented by `filter::Error::inner(failure)`.